src/fsharp/NameResolution.fs

```diff
@@ -770,57 +788,376 @@ let AddUnionCases2 bulkAddMode (eUnqualifiedItems: UnqualifiedItems) (ucrefs: Un
+        match ty with
+        | AppTy g (tcref, _) ->
+            let tycon = tcref.Deref
+            let mty = tycon.ModuleOrNamespaceType
+            // No dotting through type generators to get to a nested type!
+#if !NO_EXTENSIONTYPING
+            if checkForGenerated then
+                CheckForDirectReferenceToGeneratedType (tcref, PermitDirectReferenceToGeneratedType.No, m)
+#else
+            checkForGenerated |> ignore
+#endif
+
+            match optFilter with
+            | Some nm ->
+                LookupTypeNameInEntityMaybeHaveArity (amap, m, ad, nm, staticResInfo, tcref)
+            | None ->
+#if !NO_EXTENSIONTYPING
+                match tycon.TypeReprInfo with
+                | TProvidedTypeExtensionPoint info ->
+                    [ for nestedType in info.ProvidedType.PApplyArray((fun sty -> sty.GetNestedTypes()), "GetNestedTypes", m) do
+                        let nestedTypeName = nestedType.PUntaint((fun t -> t.Name), m)
+                        yield! 
+                            LookupTypeNameInEntityMaybeHaveArity (amap, m, ad, nestedTypeName, staticResInfo, tcref) ]
+
+                | _ ->
+#endif
+                    mty.TypesByAccessNames.Values
+                    |> List.choose (fun entity ->
+                        let tcref = tcref.NestedTyconRef entity
+                        if IsEntityAccessible amap m ad tcref then Some tcref else None)
+        | _ -> [])
+
+/// Make a type that refers to a nested type.
+///
+/// Handle the .NET/C# business where nested generic types implicitly accumulate the type parameters
+/// from their enclosing types.
+let MakeNestedType (ncenv: NameResolver) (tinst: TType list) m (tcrefNested: TyconRef) =
+    let tps = List.skip tinst.Length (tcrefNested.Typars m)
+    let tinstNested = ncenv.InstantiationGenerator m tps
+    mkAppTy tcrefNested (tinst @ tinstNested)
+
+/// Get all the accessible nested types of an existing type.
+let GetNestedTypesOfType (ad, ncenv: NameResolver, optFilter, staticResInfo, checkForGenerated, m) ty =
+    let tinst, tcrefsNested = GetNestedTyconRefsOfType ncenv.InfoReader ncenv.amap (ad, optFilter, staticResInfo, checkForGenerated, m) ty
+    tcrefsNested
+    |> List.map (MakeNestedType ncenv tinst m)
+
+let ChooseMethInfosForNameEnv g m ty (minfos: MethInfo list) =
+    let isExtTy = IsTypeUsedForCSharpStyleExtensionMembers g m ty
+
+    let methGroups =
+        minfos
+        |> List.filter (fun minfo ->
+            not (minfo.IsInstance || minfo.IsClassConstructor || minfo.IsConstructor) && typeEquiv g minfo.ApparentEnclosingType ty &&
+            not (IsMethInfoPlainCSharpStyleExtensionMember g m isExtTy minfo))
+        |> List.groupBy (fun minfo -> minfo.LogicalName)
+
+    seq {
+        for (methName, methGroup) in methGroups do
+            if not methGroup.IsEmpty then
+                KeyValuePair(methName, Item.MethodGroup(methName, methGroup, None)) 
+    }
```

Comment:
More of a style thing but may want to consider

```fsharp
|> List.filter (fun (_, methGroup) -> not methGroup.IsEmpty)
|> List.map (fun (methName, methGroup) -> KeyValuePair(...)
```

for this and others. I don't know if it has a perf impact but construction and interation of sequences is generally expensive so it might be better to do it this way if it ends up being a hot path of some sort
```

Reply:
When using `seq` I was considering perf as I didn't want to create a ton of lists all the time.

Reply:
Was that measured?